### ISSUE TITLE
lighter input tag

### ISSUE THREAD

**Issue by thipages** [WebReflection/lighterhtml#19]
Hi,
I was asking myself about the relatoinship of

- user interaction (on input , eg cheching a checkbox or entering text), something lighterhtml may not be aware of

- and ligherhtml renderer refresh. 

While trying to set-up an raw example, I wrote a preliminary one.... that I do not understand.

In the following code

- the second input checkbox is set to empty string while displayed a non-expected 'checked' checkbox

- At  the opposite the first checkbox filled alternatively with empty string and 'checked', which is ok

So, same input, different behaviour. Can you help? Thank you.

```javascript
const {render, html} = lighterhtml;
let high=false, checked;
setInterval(update,2000);

function update() {
  checked=high?"checked":"";
  high=!high;
  render(document.body, () => html`
    <input type="checkbox" checked="${checked}"/>
    <input type="text" value="${checked}"/>
    <br/>
    <input type="checkbox" checked=""/>
    <input type="text" value=""/>
`);
}
```

**Comment by WebReflection** [comment]
it's documented in here, `checked` is a boolean attribute, you don't pass the string `checked`, you pass `true` or `false` directly.

https://viperhtml.js.org/hyperhtml/documentation/#essentials-5

**Comment by WebReflection** [comment]
```js
const {render, html} = lighterhtml;
let checked = false;
setInterval(update,2000);

function update() {
  checked = !checked;
  render(document.body, () => html`
    <input type="checkbox" checked="${checked}">
    <input type="text" value="${checked}">
`);
}
```

**Comment by thipages** [comment]
Either with checked/true or empty string/false  it is the same output

Original question is that the code has  different outputs for the same input

- same input (false)

- different behaviour ('non-checked' for first checkbox 'checked' for the second)


``` javascript
const {render, html} = lighterhtml;
let checked=true;
setInterval(update,2000);

function update() {
  checked=!checked;
  render(document.body, () => html`
    <input type="checkbox" checked="${checked}"/>
    <input type="text" value="${checked}"/>
    <br/>
    <input type="checkbox" checked="false"/>
    <input type="text" value=""/>
`);
}
```

**Comment by thipages** [comment]
ok, sorry to have been a bit insiting/heavy

As a conclusion as you mention in the doc 

> special attributes

or

> Whenever an attribute has a special meaning

there is a lighterhtml rule saying that we should use a boolean map for those

seems that there is no vocabulary for naming those special attributes. Not found at MDN at least.

**Comment by WebReflection** [comment]
> Original question is that the code has different outputs for the same input

The code has always predictable output depending on the kind of nodes, and related attributes, you are dealing with.

What is your issue, exactly?

**Comment by WebReflection** [comment]
Actually, let me try again: do you understand this fails your expectations ? `<input type="checkbox" checked="false"/>` it doesn't in hyper/lighterhtml, as documented.

Any better?

**Comment by WebReflection** [comment]
P.S. the vocabulary for special attributes is here:
https://html.spec.whatwg.org/multipage/input.html#the-input-element

**Comment by thipages** [comment]
I think everything is ok now (for that topic).

My confusion was that, **for those attributes**  ,  writing in HTML (only keys, no value) is different than writing in ligherhtml  , but the [HTMLInputInterface](https://html.spec.whatwg.org/multipage/input.html#the-input-element) is clear, all those types are returning boolean...

Thank you for taking time for such explanation/guidance

**Comment by WebReflection** [comment]
next up, `<input  onchange=${handler}>` doesn't really set a handler directly on the input, it uses `addEventListener` behind the scene.

Other attribute cases should be described in the docs, let me know if you need any extra clarification 👋

**Comment by thipages** [comment]
ok, thanks for giving the way lighterhtml implements listener

I have no clear overview of HTML attributes diversity, so I may ring you on a case/case basis.

I think I got now a first understanding of how lighterhtml works which is ok for my current need
It is efficient and an pleasure to work with.

Two other topics I will need to look at are 

- the "outside lighterhtml" possible interferences (user interaction or changes via third parties)
- generic components (seems limited since all is declarative)

I keep you inform on troubles/opportunities

Thank you again.

**Comment by WebReflection** [comment]
> the "outside lighterhtml" possible interferences

lighterhtml own its own content. If you drop nodes, erase content, etc, you gonna have hard time. Obtrusive libraries are indeed not super welcome here, but that's the same with any other framework, I guess.

> generic components (seems limited since all is declarative)

you can either use Custom Elements and keep it declarative in the HTML world, or give Neverland a try (link in the README)

**Comment by thipages** [comment]
Point 1 : yep thats the current state and lighterhtml is building something in an existing middle. There might be some hard restriction.
Point 2 : custom and generic are different (may be a dream,  but let us go)

Note : currently i am not conviced about neverland (at least useState **or** for few dedicated cases cause on non persistent locals like label/input with for/id **or** I miss something) but in lighterhtml.

**Comment by WebReflection** [comment]
neverland is based on lightrhtml and dom-augmentor, it just glues these together
